### PR TITLE
Extend "use from inventory or ground" to vehicles

### PR DESCRIPTION
**Pull request by Coolthulhu**
As in name. Currently 3 functions use the inventory-or-ground menu: eating, unloading without a held item and liquid pouring (as in: picking the destination).

Now, it works already, but I'm not sure about the code quality:

I wrote a wrapper class to simplify removing the item on the caller site, without having to handle every case separately. This may look like an overkill, but if many functions start using the inventory/ground/vehicle menu, it will probably end up saving time and work.

This wrapper class is then wrapped in an `unique_ptr` to avoid bloating game.h with its header (since it would include it in 90% source files, game.h is ugly like that) while keeping the automatic deletion and the ability to use inheritance. Not sure how "proper" is this extra wrapping or the usage of a class with inheritance here. May be a bit of an overkill.

The wrapper class does some sanity checks: first at constructor, then at item retrieval. The one at the constructor may be removed as redundant or it could be used for debug messages. The code is certainly not in a performance-critical section, and barring incredibly exotic cases (a maptileful of skewers and a truckful of skewers over it, with a player with an inventory full of skewers over that), should not have a noticeable impact on performance.

### CODE REVIEW

**Comment by Rivet-the-Zombie** [comment]
Very nice!

File: src/item_location.cpp [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+void item_on_vehicle::remove_item()
+{
+    if( what == nullptr ) {
+        return;
+    }
+
+    const auto parts = veh->parts_at_relative( local_coords.x, local_coords.y );
+    for( const int i : parts ) {
+        for( item &it : veh->get_items( i ) ) {
```
Comment (BevapDin) [inline src/item_location.cpp]: This loop is completely pointless. You're not even using the loop variable `it`.

File: src/item_location.cpp [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    for( const int i : parts ) {
+        for( item &it : v.get_items( i ) ) {
+            if( &it == which ) {
+                what = &it;
+                return;
+            }
+        }
+    }
+
+    what = nullptr;
```
Comment (BevapDin) [inline src/item_location.cpp]: I think reaching this code (and similar in other functions here) is a very unexpected situation and should at least print a debug message.

File: src/item_location.cpp [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    }
+
+    auto items = who->items_with( [this]( const item &it ) {
+        return &it == what;
+    } );
+
+    if( !items.empty() ) {
+        return items[0];
+    } else {
+        what = nullptr;
```
Comment (BevapDin) [inline src/item_location.cpp]: This should also show a debug message as it indicates a very serious problem: the item somebody tried to use has vanished unexpectedly.

File: src/inventory_ui.cpp [hunk 4]
```diff
@@ -860,40 +856,88 @@ std::pair< int, item* > game::inv_map_splice( item_filter inv_filter, item_filte
...
+
+std::unique_ptr<item_location> game::inv_map_splice( item_filter filter, const std::string &title )
+{
+    return inv_map_splice( filter, filter, filter, title );
+}
+
+std::unique_ptr<item_location> game::inv_map_splice(
+    item_filter inv_filter, item_filter ground_filter, item_filter vehicle_filter, const std::string &title )
+{
+    (void)vehicle_filter;
```
Comment (BevapDin) [inline src/inventory_ui.cpp]: The parameter is actually used on line 911.

File: src/item_location.h [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+class item;
+class Character;
+class vehicle;
+
+/**
+ * A class for easy removal of used items.
+ * Ensures the item exists, but not that the character/vehicle does.
+ * Should not be kept, but removed before the end of turn.
+ */
+class item_location {
```
Comment (BevapDin) [inline src/item_location.h]: The class needs a virtual destructor. http://stackoverflow.com/a/461224

File: src/item_location.cpp [hunk 6]
```diff
@@ -0,0 +1,182 @@
+#include "item_location.h"
+
+#include "enums.h"
+#include "game.h"
+#include "map.h"
+#include "Character.h"
+#include "player.h"
+#include "vehicle.h"
+#include <map>
```
Comment (BevapDin) [inline src/item_location.cpp]: Why do you need `map` here?

**Comment by BevapDin** [comment]
Nice.

I'm a bit worried by the additional indirection. It might be useful to change functions (in the player class for example) to accept a `item_location` reference. The caller can get that either from the inventory functions or just create one on the stack.

This has actually the potential to replace most of the existing item references.

And the new files need to be added to CataclysmWin.cbp and src/CMaleLists.txt and the header is named character.h (lowercase c).

**Comment by Coolthulhu** [comment]
Cleaned up

What do you mean by indirection here? The extra `unique_ptr` wrapper, the fact that it's game class and not player class which handles the item removal? Or something else altogether?

**Comment by BevapDin** [comment]
> What do you mean by indirection here? The extra  unique_ptr  wrapper, the fact that it's game class and not player class which handles the item removal? Or something else altogether?

Basically the unique_ptr, under the assumption that most function will use the item object directly and do not care where it comes from. The unique ptr adds another point of failure, it can contain a `nullptr`. Now you have (theoretically) to check for containing a nullptr.

What about making `item_location` a wrapper that contains the unique_ptr itself? The few subclases can than be moved into the cpp file (their structure does not need to be exposed), they are created via 4 different constructors. Something like the Pimpl idiom.

```C++
class item_location {
private:
    class impl;
    std::unique_ptr<impl> ptr;

public:
    item_location(); // : ptr(new item_is_null())
    item_location(Character&p, int i); // : ptr(new item_on_person(p, i))
    item_location(tripoint &p); // : ptr(new item_on_map(p))

    void remove_item();
    item *get_item();
};
```

The item_location class would not have any virtual functions, and the whole implementation would be hidden in the cpp file. This would be a bit easier to use, I think. The public interface is quite small and if someone comes across this class, they can easily understand it. It also makes sure that the `unique_ptr` is never NULL because nobody but `item_location` can access (write to) it.

Also a nice thing: because the classes are declared *and* defined in the cpp file, the functions can be put into the class declaration:

```C++
// content of cpp file:
class item_on_person : public item_location::impl {
public:
    void remove_item() {
        // implementation right here, no separate definition of the function.
    }
    ...
};
```

**Comment by Coolthulhu** [comment]
OK, now it does look far cleaner than my version.

I couldn't get the implementation to be a nested class without also forward declaring all its derived classes in the header. Compiler would complain about trying to derive from a private class. At the moment the implementation is a separate, non-nested class.
Any good trick for that?

**Comment by Coolthulhu** [comment]
Couldn't get around the above, so I forward declared all the derived classes of `impl` as nested classes of `item_location`.

Forgot to mention before: I decided to create `item_location` objects with static methods rather than explicit constructor. Felt it would be more readable this way.

I think it should be ready now.

**Comment by BevapDin** [comment]
Alternatively you could have put `impl` into a namespace, that makes it available outside the `item_location` class but does not clutter up the global namespace.

I think you don't need to include "item_location.h" in game.h, a forward declaration of `item_location` should work. Only code that calls those function / defines them needs to include the header.

**Pull request merged**